In [1]:
from config import CONFIG
from modes import build_runtime_mode_by_name
from workloads import build_runtime_workload_by_name
from model_loader import load_model_for_mode, unload_model
from runner import run_single_benchmark

In [2]:
from huggingface_hub import login
login(token="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")

In [3]:
from benchmark_modes import (
    _timestamp_str,
    run_full_benchmark,
    save_results_json,
    save_results_csv,
    save_summary_csv,
    build_aggregate_rows,
    build_comparison_rows,
    save_aggregate_csv,
    save_test_table_md,
)
from config import RAW_RESULTS_DIR, CONFIG

# Increase this variable for longer runs
CONFIG.system.num_trials = 2

timestamp = _timestamp_str()

results = run_full_benchmark(
    include_hybrids=True,
    repeated_prefix_variants=2,
    test_profile="controller",
)

Starting curated large benchmark run
Test profile: controller
Enabled modes: 11
Curated test cases: 42
Trials per pair: 2
Total runs: 84
Per-run logs will be saved under: /scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/logs


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[1/84] fp16_baseline            | short_prompt_short_output    | trial=0 | ttft=29.52 ms | lat=708.35 ms | tps=43.76 | J/tok=3.989 | gpu=32663.21 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[2/84] fp16_baseline            | short_prompt_short_output    | trial=1 | ttft=25.49 ms | lat=733.62 ms | tps=42.26 | J/tok=4.404 | gpu=32855.21 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[3/84] fp16_baseline            | short_prompt_long_output     | trial=0 | ttft=25.19 ms | lat=2760.21 ms | tps=46.01 | J/tok=4.248 | gpu=32855.21 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[4/84] fp16_baseline            | short_prompt_long_output     | trial=1 | ttft=24.97 ms | lat=2746.45 ms | tps=46.24 | J/tok=4.151 | gpu=32855.21 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[5/84] fp16_baseline            | long_prompt_short_output     | trial=0 | ttft=109.32 ms | lat=783.06 ms | tps=40.87 | J/tok=5.393 | gpu=32992.67 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[6/84] fp16_baseline            | long_prompt_short_output     | trial=1 | ttft=108.22 ms | lat=775.41 ms | tps=41.27 | J/tok=5.244 | gpu=32992.67 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[7/84] fp16_baseline            | long_prompt_long_output      | trial=0 | ttft=107.69 ms | lat=2796.98 ms | tps=45.76 | J/tok=4.514 | gpu=32992.67 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[8/84] fp16_baseline            | long_prompt_long_output      | trial=1 | ttft=108.00 ms | lat=2826.30 ms | tps=45.29 | J/tok=4.483 | gpu=32992.67 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[9/84] fp16_baseline            | memory_pressure_long_context | trial=0 | ttft=173.79 ms | lat=2973.38 ms | tps=43.05 | J/tok=4.937 | gpu=33103.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[10/84] fp16_baseline            | memory_pressure_long_context | trial=1 | ttft=175.79 ms | lat=2996.58 ms | tps=42.72 | J/tok=4.772 | gpu=33103.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[11/84] fp16_baseline            | shared_prefix_chat_v0        | trial=0 | ttft=101.88 ms | lat=1455.29 ms | tps=43.98 | J/tok=4.963 | gpu=32989.42 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[12/84] fp16_baseline            | shared_prefix_chat_v0        | trial=1 | ttft=101.26 ms | lat=1470.82 ms | tps=43.51 | J/tok=5.266 | gpu=32989.42 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[13/84] fp16_baseline            | shared_prefix_chat_v1        | trial=0 | ttft=105.51 ms | lat=1460.22 ms | tps=43.83 | J/tok=5.185 | gpu=32984.71 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[14/84] fp16_baseline            | shared_prefix_chat_v1        | trial=1 | ttft=104.10 ms | lat=1485.68 ms | tps=43.08 | J/tok=5.350 | gpu=32984.71 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[15/84] int8_quant               | short_prompt_short_output    | trial=0 | ttft=37.57 ms | lat=377.66 ms | tps=82.09 | J/tok=1.840 | gpu=32828.84 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[16/84] int8_quant               | short_prompt_short_output    | trial=1 | ttft=37.87 ms | lat=376.89 ms | tps=82.25 | J/tok=1.719 | gpu=32892.84 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[17/84] int8_quant               | short_prompt_long_output     | trial=0 | ttft=43.86 ms | lat=1416.28 ms | tps=89.67 | J/tok=2.124 | gpu=32892.84 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[18/84] int8_quant               | short_prompt_long_output     | trial=1 | ttft=37.27 ms | lat=1413.51 ms | tps=89.85 | J/tok=2.122 | gpu=32892.84 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[19/84] int8_quant               | long_prompt_short_output     | trial=0 | ttft=100.76 ms | lat=446.79 ms | tps=71.62 | J/tok=2.169 | gpu=33029.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[20/84] int8_quant               | long_prompt_short_output     | trial=1 | ttft=102.52 ms | lat=446.42 ms | tps=71.68 | J/tok=2.559 | gpu=33029.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[21/84] int8_quant               | long_prompt_long_output      | trial=0 | ttft=99.38 ms | lat=1493.14 ms | tps=85.73 | J/tok=2.306 | gpu=33029.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[22/84] int8_quant               | long_prompt_long_output      | trial=1 | ttft=103.04 ms | lat=1498.72 ms | tps=85.41 | J/tok=2.286 | gpu=33029.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[23/84] int8_quant               | memory_pressure_long_context | trial=0 | ttft=150.23 ms | lat=1569.75 ms | tps=81.54 | J/tok=2.479 | gpu=33140.91 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[24/84] int8_quant               | memory_pressure_long_context | trial=1 | ttft=156.46 ms | lat=1570.16 ms | tps=81.52 | J/tok=2.475 | gpu=33140.91 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[25/84] awq_4bit                 | short_prompt_short_output    | trial=0 | ttft=75.13 ms | lat=401.82 ms | tps=77.15 | J/tok=2.441 | gpu=32850.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[26/84] awq_4bit                 | short_prompt_short_output    | trial=1 | ttft=73.84 ms | lat=402.26 ms | tps=77.06 | J/tok=2.313 | gpu=32850.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[27/84] awq_4bit                 | short_prompt_long_output     | trial=0 | ttft=71.49 ms | lat=1415.85 ms | tps=89.70 | J/tok=2.386 | gpu=32850.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[28/84] awq_4bit                 | short_prompt_long_output     | trial=1 | ttft=74.67 ms | lat=1415.88 ms | tps=89.70 | J/tok=2.350 | gpu=32850.65 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[29/84] awq_4bit                 | long_prompt_short_output     | trial=0 | ttft=156.99 ms | lat=490.79 ms | tps=65.20 | J/tok=3.873 | gpu=33101.32 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[30/84] awq_4bit                 | long_prompt_short_output     | trial=1 | ttft=173.78 ms | lat=507.71 ms | tps=63.03 | J/tok=3.374 | gpu=33101.32 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[31/84] awq_4bit                 | long_prompt_long_output      | trial=0 | ttft=168.69 ms | lat=1532.96 ms | tps=83.50 | J/tok=2.657 | gpu=33101.32 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[32/84] awq_4bit                 | long_prompt_long_output      | trial=1 | ttft=160.66 ms | lat=1528.14 ms | tps=83.76 | J/tok=2.644 | gpu=33101.32 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[33/84] awq_4bit                 | memory_pressure_long_context | trial=0 | ttft=249.28 ms | lat=1636.30 ms | tps=78.23 | J/tok=2.960 | gpu=33184.25 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[34/84] awq_4bit                 | memory_pressure_long_context | trial=1 | ttft=244.35 ms | lat=1634.31 ms | tps=78.32 | J/tok=2.840 | gpu=33184.25 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[35/84] gptq_4bit                | short_prompt_short_output    | trial=0 | ttft=51.63 ms | lat=279.24 ms | tps=114.60 | J/tok=1.040 | gpu=33008.40 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[36/84] gptq_4bit                | short_prompt_short_output    | trial=1 | ttft=51.73 ms | lat=282.39 ms | tps=113.32 | J/tok=1.260 | gpu=33018.33 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[37/84] gptq_4bit                | short_prompt_long_output     | trial=0 | ttft=51.49 ms | lat=950.31 ms | tps=134.69 | J/tok=1.450 | gpu=33026.46 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[38/84] gptq_4bit                | short_prompt_long_output     | trial=1 | ttft=51.50 ms | lat=941.34 ms | tps=135.98 | J/tok=1.495 | gpu=33034.58 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[39/84] gptq_4bit                | long_prompt_short_output     | trial=0 | ttft=146.64 ms | lat=369.96 ms | tps=86.50 | J/tok=2.412 | gpu=33142.26 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[40/84] gptq_4bit                | long_prompt_short_output     | trial=1 | ttft=139.72 ms | lat=363.75 ms | tps=87.97 | J/tok=2.475 | gpu=33150.39 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[41/84] gptq_4bit                | long_prompt_long_output      | trial=0 | ttft=149.69 ms | lat=1063.71 ms | tps=120.33 | J/tok=1.755 | gpu=33158.51 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[42/84] gptq_4bit                | long_prompt_long_output      | trial=1 | ttft=146.35 ms | lat=1063.01 ms | tps=120.41 | J/tok=1.858 | gpu=33166.64 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[43/84] gptq_4bit                | memory_pressure_long_context | trial=0 | ttft=216.37 ms | lat=1150.90 ms | tps=111.22 | J/tok=2.037 | gpu=33257.60 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[44/84] gptq_4bit                | memory_pressure_long_context | trial=1 | ttft=218.46 ms | lat=1154.83 ms | tps=110.84 | J/tok=1.989 | gpu=33265.72 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[45/84] speculative_decoding     | short_prompt_short_output    | trial=0 | ttft=42.60 ms | lat=339.64 ms | tps=91.27 | J/tok=1.391 | gpu=32965.44 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[46/84] speculative_decoding     | short_prompt_short_output    | trial=1 | ttft=41.28 ms | lat=333.04 ms | tps=93.08 | J/tok=1.659 | gpu=33021.69 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[47/84] speculative_decoding     | short_prompt_long_output     | trial=0 | ttft=41.04 ms | lat=1284.71 ms | tps=98.86 | J/tok=2.057 | gpu=33037.94 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[48/84] speculative_decoding     | short_prompt_long_output     | trial=1 | ttft=40.49 ms | lat=1270.60 ms | tps=99.95 | J/tok=2.071 | gpu=33054.19 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[49/84] speculative_decoding     | long_prompt_short_output     | trial=0 | ttft=153.89 ms | lat=414.02 ms | tps=77.29 | J/tok=2.417 | gpu=33204.45 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


Loading safetensors checkpoint shards:   0% Completed | 0/1 [00:00<?, ?it/s]


[50/84] speculative_decoding     | long_prompt_short_output     | trial=1 | ttft=144.98 ms | lat=402.35 ms | tps=79.53 | J/tok=2.618 | gpu=33220.70 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[51/84] kv_cache_compression     | short_prompt_long_output     | trial=0 | ttft=54.76 ms | lat=2317.29 ms | tps=54.81 | J/tok=3.941 | gpu=33090.09 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[52/84] kv_cache_compression     | short_prompt_long_output     | trial=1 | ttft=55.15 ms | lat=2334.01 ms | tps=54.41 | J/tok=4.009 | gpu=33098.22 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[53/84] kv_cache_compression     | long_prompt_long_output      | trial=0 | ttft=133.65 ms | lat=2755.49 ms | tps=46.45 | J/tok=4.737 | gpu=33243.81 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[54/84] kv_cache_compression     | long_prompt_long_output      | trial=1 | ttft=137.90 ms | lat=2755.20 ms | tps=46.46 | J/tok=4.700 | gpu=33251.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[55/84] kv_cache_compression     | memory_pressure_long_context | trial=0 | ttft=200.53 ms | lat=2836.57 ms | tps=45.12 | J/tok=5.160 | gpu=33362.91 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[56/84] kv_cache_compression     | memory_pressure_long_context | trial=1 | ttft=214.47 ms | lat=2840.21 ms | tps=45.07 | J/tok=5.047 | gpu=33362.91 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[57/84] prefix_caching           | shared_prefix_chat_v0        | trial=0 | ttft=119.85 ms | lat=1057.59 ms | tps=60.51 | J/tok=4.364 | gpu=33248.68 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[58/84] prefix_caching           | shared_prefix_chat_v0        | trial=1 | ttft=117.85 ms | lat=1056.14 ms | tps=60.60 | J/tok=4.305 | gpu=33248.68 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[59/84] prefix_caching           | shared_prefix_chat_v1        | trial=0 | ttft=121.01 ms | lat=1057.00 ms | tps=60.55 | J/tok=4.391 | gpu=33240.39 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[60/84] prefix_caching           | shared_prefix_chat_v1        | trial=1 | ttft=121.08 ms | lat=1060.98 ms | tps=60.32 | J/tok=4.589 | gpu=33240.39 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[61/84] chunked_prefill          | long_prompt_short_output     | trial=0 | ttft=142.07 ms | lat=599.82 ms | tps=53.35 | J/tok=4.201 | gpu=33251.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[62/84] chunked_prefill          | long_prompt_short_output     | trial=1 | ttft=126.01 ms | lat=583.68 ms | tps=54.82 | J/tok=4.528 | gpu=33251.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[63/84] chunked_prefill          | long_prompt_long_output      | trial=0 | ttft=123.37 ms | lat=1999.75 ms | tps=64.01 | J/tok=3.967 | gpu=33251.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[64/84] chunked_prefill          | long_prompt_long_output      | trial=1 | ttft=131.75 ms | lat=2007.46 ms | tps=63.76 | J/tok=3.907 | gpu=33251.93 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[65/84] chunked_prefill          | memory_pressure_long_context | trial=0 | ttft=193.87 ms | lat=2088.62 ms | tps=61.28 | J/tok=4.268 | gpu=33371.03 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[66/84] chunked_prefill          | memory_pressure_long_context | trial=1 | ttft=192.01 ms | lat=2081.24 ms | tps=61.50 | J/tok=4.084 | gpu=33371.03 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[67/84] continuous_batching      | short_prompt_short_output    | trial=0 | ttft=86.37 ms | lat=555.32 ms | tps=230.50 | J/tok=0.910 | gpu=33184.55 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[68/84] continuous_batching      | short_prompt_short_output    | trial=1 | ttft=73.89 ms | lat=541.99 ms | tps=236.17 | J/tok=0.939 | gpu=33192.68 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[69/84] continuous_batching      | short_prompt_long_output     | trial=0 | ttft=86.29 ms | lat=2011.12 ms | tps=254.58 | J/tok=0.939 | gpu=33192.68 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[70/84] continuous_batching      | short_prompt_long_output     | trial=1 | ttft=86.07 ms | lat=2003.76 ms | tps=255.52 | J/tok=0.916 | gpu=33192.68 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[71/84] continuous_batching      | long_prompt_short_output     | trial=0 | ttft=432.44 ms | lat=933.88 ms | tps=137.06 | J/tok=2.032 | gpu=33435.24 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[72/84] continuous_batching      | long_prompt_short_output     | trial=1 | ttft=422.41 ms | lat=919.74 ms | tps=139.17 | J/tok=2.090 | gpu=33435.24 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[73/84] cuda_graphs              | short_prompt_short_output    | trial=0 | ttft=40.37 ms | lat=493.17 ms | tps=62.86 | J/tok=3.458 | gpu=33138.84 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[74/84] cuda_graphs              | short_prompt_short_output    | trial=1 | ttft=44.03 ms | lat=495.41 ms | tps=62.57 | J/tok=3.334 | gpu=33146.96 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[75/84] cuda_graphs              | short_prompt_long_output     | trial=0 | ttft=41.33 ms | lat=1896.30 ms | tps=66.97 | J/tok=3.650 | gpu=33146.96 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[76/84] cuda_graphs              | short_prompt_long_output     | trial=1 | ttft=42.26 ms | lat=1895.58 ms | tps=67.00 | J/tok=3.679 | gpu=33146.96 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[77/84] cuda_graphs              | long_prompt_short_output     | trial=0 | ttft=146.21 ms | lat=603.70 ms | tps=53.01 | J/tok=4.421 | gpu=33292.55 MB


Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]


[78/84] cuda_graphs              | long_prompt_short_output     | trial=1 | ttft=127.25 ms | lat=586.86 ms | tps=54.53 | J/tok=4.830 | gpu=33292.55 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[79/84] awq_plus_chunked_prefill | long_prompt_short_output     | trial=0 | ttft=159.03 ms | lat=494.39 ms | tps=64.73 | J/tok=3.561 | gpu=33369.20 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[80/84] awq_plus_chunked_prefill | long_prompt_short_output     | trial=1 | ttft=173.81 ms | lat=507.31 ms | tps=63.08 | J/tok=3.660 | gpu=33369.20 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[81/84] awq_plus_chunked_prefill | long_prompt_long_output      | trial=0 | ttft=180.33 ms | lat=1546.79 ms | tps=82.75 | J/tok=2.570 | gpu=33369.20 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[82/84] awq_plus_chunked_prefill | long_prompt_long_output      | trial=1 | ttft=170.26 ms | lat=1536.81 ms | tps=83.29 | J/tok=2.614 | gpu=33369.20 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[83/84] awq_plus_chunked_prefill | memory_pressure_long_context | trial=0 | ttft=228.41 ms | lat=1620.18 ms | tps=79.00 | J/tok=3.070 | gpu=33452.12 MB


Loading safetensors checkpoint shards:   0% Completed | 0/2 [00:00<?, ?it/s]


[84/84] awq_plus_chunked_prefill | memory_pressure_long_context | trial=1 | ttft=226.81 ms | lat=1617.95 ms | tps=79.11 | J/tok=2.984 | gpu=33452.12 MB

Benchmark sweep complete.
Collected results: 84


In [4]:
json_path = RAW_RESULTS_DIR / f"benchmark_results_{timestamp}.json"
csv_path = RAW_RESULTS_DIR / f"benchmark_results_{timestamp}.csv"
summary_csv_path = RAW_RESULTS_DIR / f"benchmark_summary_{timestamp}.csv"
aggregate_csv_path = RAW_RESULTS_DIR / f"benchmark_aggregates_{timestamp}.csv"
comparison_csv_path = RAW_RESULTS_DIR / f"benchmark_comparisons_{timestamp}.csv"
table_md_path = RAW_RESULTS_DIR / f"benchmark_table_{timestamp}.md"
comparison_md_path = RAW_RESULTS_DIR / f"benchmark_comparisons_{timestamp}.md"

save_results_json(results, json_path)
save_results_csv(results, csv_path)
save_summary_csv(results, summary_csv_path)

aggregate_rows = build_aggregate_rows(results)
comparison_rows = build_comparison_rows(results)

save_aggregate_csv(aggregate_rows, aggregate_csv_path)
save_aggregate_csv(comparison_rows, comparison_csv_path)

save_test_table_md(
    rows=aggregate_rows,
    output_path=table_md_path,
    title="Benchmark Table",
    columns=[
        "mode_name",
        "workload_name",
        "completed_runs",
        "failed_runs",
        "ttft_ms_mean",
        "total_latency_ms_mean",
        "tokens_per_second_mean",
        "avg_power_w_mean",
        "energy_per_token_j_mean",
        "peak_gpu_memory_mb_mean",
        "reference_rouge_l_f1_mean",
        "failure_rate",
    ],
)

save_test_table_md(
    rows=comparison_rows,
    output_path=comparison_md_path,
    title="Benchmark Comparisons vs FP16 Baseline",
    columns=[
        "mode_name",
        "workload_name",
        "latency_speedup_vs_baseline",
        "throughput_ratio_vs_baseline",
        "energy_per_token_ratio_vs_baseline",
        "peak_gpu_memory_delta_vs_baseline_mb",
        "quality_degradation_vs_baseline_mean",
        "failure_rate",
    ],
)

print("Saved:")
print(json_path)
print(csv_path)
print(summary_csv_path)
print(aggregate_csv_path)
print(comparison_csv_path)
print(table_md_path)
print(comparison_md_path)

Saved:
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_results_20260422_222918.json
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_results_20260422_222918.csv
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_summary_20260422_222918.csv
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_aggregates_20260422_222918.csv
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_comparisons_20260422_222918.csv
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_table_20260422_222918.md
/scratch/as18181/ModeSwitch-LLM/ModeSwitch-LLM/results/raw/benchmark_comparisons_20260422_222918.md


In [8]:
import pandas as pd
from IPython.display import display

agg_df = pd.read_csv(aggregate_csv_path)
cmp_df = pd.read_csv(comparison_csv_path)

agg_show = agg_df.sort_values(
    ["workload_name", "tokens_per_second_mean"],
    ascending=[True, False]
)

cmp_show = cmp_df.sort_values(
    ["workload_name", "latency_speedup_vs_baseline"],
    ascending=[True, False]
)

print("AGGREGATE TABLE")
display(agg_show)

print("COMPARISON VS BASELINE")
display(cmp_show)

AGGREGATE TABLE


,mode_name,workload_name,backend,num_runs,completed_runs,failed_runs,failure_rate,most_common_error_type,reference_rouge_l_f1_mean,baseline_similarity_rouge_l_f1_mean,...,total_latency_ms_mean,total_latency_ms_std,total_latency_ms_p50,total_latency_ms_p95,total_latency_ms_p99,tokens_per_second_mean,tokens_per_second_std,tokens_per_second_p50,tokens_per_second_p95,tokens_per_second_p99
20,gptq_4bit,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.260870,0.228814,...,1063.358243,0.352831,1063.358243,1063.675791,1063.704017,120.373369,0.039941,120.373369,120.409316,120.412511
10,int8_quant,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.299465,0.836820,...,1495.930020,2.787608,1495.930020,1498.438867,1498.661876,85.565797,0.159449,85.565797,85.709301,85.722057
15,awq_4bit,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.279570,0.210084,...,1530.549567,2.414266,1530.549567,1532.722406,1532.915548,83.630299,0.131917,83.630299,83.749025,83.759578
40,awq_plus_chunked_prefill,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.279570,0.210084,...,1541.799426,4.986538,1541.799426,1546.287311,1546.686234,83.020746,0.268508,83.020746,83.262404,83.283884
31,chunked_prefill,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.293478,1.000000,...,2003.607287,3.853159,2003.607287,2007.075130,2007.383383,63.885011,0.122858,63.885011,63.995583,64.005412
26,kv_cache_compression,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.265957,0.700000,...,2755.344086,0.144852,2755.344086,2755.474453,2755.486041,46.455178,0.002442,46.455178,46.457376,46.457572
3,fp16_baseline,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.293478,1.000000,...,2811.639575,14.659986,2811.639575,2824.833562,2826.006361,45.526276,0.237376,45.526276,45.739914,45.758904
35,continuous_batching,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.180952,0.315789,...,926.810897,7.070961,926.810897,933.174762,933.740439,138.116040,1.053735,138.116040,139.064402,139.148700
19,gptq_4bit,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.150538,0.185185,...,366.855012,3.105325,366.855012,369.649805,369.898231,87.234171,0.738413,87.234171,87.898743,87.957816
24,speculative_decoding,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.236559,1.000000,...,408.183085,5.832413,408.183085,413.432257,413.898850,78.412202,1.120410,78.412202,79.420571,79.510204


COMPARISON VS BASELINE


,mode_name,workload_name,backend,num_runs,completed_runs,failed_runs,failure_rate,most_common_error_type,reference_rouge_l_f1_mean,baseline_similarity_rouge_l_f1_mean,...,total_latency_ms_p99,tokens_per_second_mean,tokens_per_second_std,tokens_per_second_p50,tokens_per_second_p95,tokens_per_second_p99,latency_speedup_vs_baseline,throughput_ratio_vs_baseline,energy_per_token_ratio_vs_baseline,peak_gpu_memory_delta_vs_baseline_mb
13,gptq_4bit,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.260870,0.228814,...,1063.704017,120.373369,0.039941,120.373369,120.409316,120.412511,2.644113,2.644042,0.401616,169.901367
3,int8_quant,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.299465,0.836820,...,1498.661876,85.565797,0.159449,85.565797,85.709301,85.722057,1.879526,1.879482,0.510365,37.260742
8,awq_4bit,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.279570,0.210084,...,1532.915548,83.630299,0.131917,83.630299,83.749025,83.759578,1.837013,1.836968,0.589293,108.651367
33,awq_plus_chunked_prefill,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.279570,0.210084,...,1546.686234,83.020746,0.268508,83.020746,83.262404,83.283884,1.823609,1.823579,0.576173,376.526855
24,chunked_prefill,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.293478,1.000000,...,2007.383383,63.885011,0.122858,63.885011,63.995583,64.005412,1.403289,1.403256,0.875258,259.261230
19,kv_cache_compression,long_prompt_long_output,vllm,2,2,0,0.0,NaN,0.265957,0.700000,...,2755.486041,46.455178,0.002442,46.455178,46.457376,46.457572,1.020431,1.020404,1.048956,255.198730
12,gptq_4bit,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.150538,0.185185,...,369.898231,87.234171,0.738413,87.234171,87.898743,87.957816,2.124098,2.124199,0.459429,153.651367
17,speculative_decoding,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.236559,1.000000,...,413.898850,78.412202,1.120410,78.412202,79.420571,79.510204,1.909036,1.909379,0.473406,219.903809
2,int8_quant,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.236559,1.000000,...,446.786362,71.651921,0.029921,71.651921,71.678850,71.681244,1.744805,1.744763,0.444543,37.260742
7,awq_4bit,long_prompt_short_output,vllm,2,2,0,0.0,NaN,0.129032,0.148148,...,507.542579,64.114190,1.086302,64.114190,65.091861,65.178766,1.560805,1.561215,0.681294,108.651367


In [ ]:
cols_agg = [
    "mode_name",
    "workload_name",
    "completed_runs",
    "failed_runs",
    "ttft_ms_mean",
    "total_latency_ms_mean",
    "tokens_per_second_mean",
    "avg_power_w_mean",
    "energy_per_token_j_mean",
    "peak_gpu_memory_mb_mean",
    "failure_rate",
]

cols_cmp = [
    "mode_name",
    "workload_name",
    "latency_speedup_vs_baseline",
    "throughput_ratio_vs_baseline",
    "energy_per_token_ratio_vs_baseline",
    "peak_gpu_memory_delta_vs_baseline_mb",
    "quality_degradation_vs_baseline_mean",
    "failure_rate",
]

print("TABLE: AGGREGATES")
display(agg_show[cols_agg])

print("TABLE: VS BASELINE")
display(cmp_show[cols_cmp])

TABLE: AGGREGATES


## Result Summary and Final Mode Selection

### What these results show
Our benchmark sweep produced a clear pattern across latency, throughput, energy per token, memory usage, and output-quality similarity to the FP16 baseline. The main takeaway is that there is **no single universal winner** across all workload types. Instead, different modes are strongest in different situations, which supports the core project idea: a **simple controller that switches modes based on workload characteristics** is more justified than always using one fixed mode.

### Strong overall findings
- **FP16 baseline** is stable and serves as the main reference, but it is consistently slower and less energy-efficient than several optimized modes.
- **INT8 quantization** is the strongest all-around practical candidate. It gives large latency and throughput improvements over FP16 across all major workload groups while keeping memory overhead small and preserving quality better than more aggressive 4-bit modes.
- **Speculative decoding** is very strong for short interactive workloads. It improves latency and throughput substantially relative to FP16 while keeping output quality effectively unchanged in the tested cases.
- **Prefix caching** clearly helps on shared-prefix workloads and should be treated as a conditional mode rather than a universal one.
- **Chunked prefill** helps on long-prompt and memory-heavy workloads, especially when prefill cost is high.
- **More aggressive 4-bit modes** like AWQ and GPTQ are very fast, but they introduce larger quality degradation than INT8 and therefore are less safe as default controller choices unless quality constraints are relaxed.

### Mode-by-mode interpretation

#### 1. FP16 baseline
FP16 is the reference point. It is not the best mode in efficiency, but it remains important because it provides the quality anchor and a stable fallback. We should **keep FP16 as the baseline and safe fallback**, not as the main optimized mode. 

#### 2. INT8 quantization
INT8 is the most balanced choice in the current results. It consistently improves latency and throughput by about **1.7× to 1.95×** over FP16 across the main workloads, while also reducing energy per token to roughly **0.42× to 0.51×** of baseline. Its memory overhead relative to FP16 is small compared with other optimized modes. This makes INT8 the **best general-purpose controller candidate**. 

#### 3. Speculative decoding
Speculative decoding performs especially well on short interactive workloads. On short-prompt workloads, it gives around **2.14× to 2.16×** latency/throughput improvement over baseline with **no measured quality degradation** in the comparison table. This makes it a strong candidate for the “fast interactive request” branch of the controller. Its current downside is that it was not benchmarked across as many workload categories as INT8. 

#### 4. Prefix caching
Prefix caching is not meant to be universally enabled; it is useful when repeated shared prefixes exist. In the shared-prefix workloads, it improves latency and throughput by about **1.38× to 1.39×** over baseline with **no measured quality degradation**. This makes it an excellent conditional mode: if shared-prefix structure is detected, prefix caching should be preferred. 

#### 5. Chunked prefill
Chunked prefill is helpful on long-prompt and memory-pressure workloads. It gives around **1.32× to 1.43×** speedup over FP16 with **no measured quality degradation** in the tested long-prompt and memory-heavy settings. This makes it a strong controller candidate for requests with large input contexts. 

#### 6. AWQ 4-bit
AWQ improves speed and energy relative to FP16, but it shows noticeably larger quality degradation than INT8 across several workloads. It is not a bad mode, but based on current results it is harder to justify as a primary controller choice when INT8 gives a better quality-efficiency tradeoff. AWQ is better treated as an aggressive optimization mode rather than a default recommended mode. 

#### 7. GPTQ 4-bit
GPTQ is the fastest and most energy-efficient mode in many tested workloads. For example, it reaches around **2.57× to 2.91×** speedup over FP16 on short-prompt workloads and **2.59× to 2.64×** on longer ones. However, its measured quality degradation is also among the highest, often much worse than INT8 and speculative decoding. So GPTQ looks strong as a “maximum-efficiency, lower-quality-tolerance” mode, but risky as a general-purpose controller option. 

#### 8. KV-cache compression
KV-cache compression underperformed relative to the other candidates. Its speedup over baseline is small, sometimes close to negligible, while memory overhead and energy efficiency are not compelling enough to justify keeping it in the first controller version. Based on current data, this mode should be removed from the main set. 

#### 9. Continuous batching
Continuous batching shows extremely high throughput, but the current evaluation is mostly single-request/controller-oriented rather than many-request serving-oriented. In that context, it is not the cleanest fit for the first controller version. It may still be valuable as a traffic-aware serving extension, but it should not be part of the initial small mode set for the controller. 

#### 10. CUDA graphs
CUDA graphs improve over baseline, but they do not stand out enough relative to INT8, speculative decoding, prefix caching, or chunked prefill. Since we want a small and interpretable controller mode library, CUDA graphs can be dropped from the first version without losing the main story. 

#### 11. AWQ + chunked prefill hybrid
The AWQ + chunked prefill hybrid does work and gives meaningful gains over baseline, but it does not clearly dominate INT8 or the cleaner specialized modes. It also inherits the quality concerns of AWQ. So while it is a useful proof-of-concept hybrid, it should not be one of the final core controller modes unless we specifically want one hybrid example in the writeup. 

### Modes to keep for the controller
Based on the current benchmark evidence, the best final controller set is:

1. **FP16 baseline** — keep as the safe fallback and reference.
2. **INT8 quantization** — keep as the best all-around efficiency mode.
3. **Speculative decoding** — keep for short, interactive, latency-sensitive requests.
4. **Prefix caching** — keep for repeated shared-prefix workloads.
5. **Chunked prefill** — keep for long-prompt / prefill-heavy / memory-heavy workloads.

This set is small, interpretable, and covers distinct workload regimes instead of containing several overlapping modes.

### Modes to cut from the first controller version
The following should be cut from the first controller version:

- **KV-cache compression** — weak gains relative to alternatives.
- **CUDA graphs** — useful, but not distinctive enough for the first small mode library.
- **Continuous batching** — better framed as a traffic-aware serving mode than a per-request controller mode.
- **AWQ 4-bit** — decent efficiency, but worse quality tradeoff than INT8.
- **GPTQ 4-bit** — very fast, but quality degradation is high enough that it is risky for the first main controller.
- **AWQ + chunked prefill hybrid** — works, but does not clearly beat simpler choices strongly enough.

### Final ranking for controller selection
If we rank the modes specifically for the **first version of the controller**, the order is:

1. **INT8 quantization**
2. **Speculative decoding**
3. **Chunked prefill**
4. **Prefix caching**
5. **FP16 baseline**
6. **GPTQ 4-bit**
7. **AWQ 4-bit**
8. **AWQ + chunked prefill**
9. **CUDA graphs**
10. **Continuous batching**
11. **KV-cache compression**

This ranking is not simply “fastest to slowest.” It prioritizes **practical controller value**, meaning a combination of:
- clear speed/energy benefit,
- manageable quality degradation,
- interpretable workload specialization,
- and usefulness in a simple rule-based controller.

### Recommended controller logic
A simple final controller can be framed as:

- **If request is short and interactive:** use **speculative decoding**
- **If request is long / prefill-heavy / memory-heavy:** use **chunked prefill**
- **If shared prefix is detected:** use **prefix caching**
- **Otherwise:** use **INT8 quantization**
- **Fallback:** use **FP16 baseline**

### Final takeaway
The experiments support the main project hypothesis: different inference modes are best for different workload regimes, so a small workload-aware selector is more justified than a single fixed mode. INT8 appears to be the strongest default optimized mode, while speculative decoding, prefix caching, and chunked prefill provide specialized gains in situations where they match the workload structure. 